controller/deployments_urlprovider.go

```diff
@@ -0,0 +1,306 @@
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/fabric8-services/fabric8-wit/app"
+	"github.com/fabric8-services/fabric8-wit/auth"
+	"github.com/fabric8-services/fabric8-wit/auth/authservice"
+	"github.com/fabric8-services/fabric8-wit/configuration"
+	"github.com/fabric8-services/fabric8-wit/kubernetes"
+	"github.com/fabric8-services/fabric8-wit/log"
+
+	goajwt "github.com/goadesign/goa/middleware/security/jwt"
+	errs "github.com/pkg/errors"
+)
+
+// there are several concrete instantiations:
+//
+//   - access /api/user/services instead of Auth
+//   - use proxy (if present) for normal OSO API calls
+//   - access OSO metrics directly (until proxy supports this)
+
+type tenantURLProvider struct {
+	apiURL     string
+	apiToken   string
+	tenant     *app.UserService
+	namespaces map[string]*app.NamespaceAttributes
+	tokens     map[string]string
+	TokenRetriever
+}
+
+// TokenRetriever is a service that will respond with an authorization token, given a service endpoint or name
+type TokenRetriever interface {
+	TokenForService(serviceURL string) (*string, error)
+}
+
+type tokenRetriever struct {
+	authClient *authservice.Client
+	context    context.Context
+}
+
+func (tr *tokenRetriever) TokenForService(forService string) (*string, error) {
+
+	resp, err := tr.authClient.RetrieveToken(tr.context, authservice.RetrieveTokenPath(), forService, nil)
+	if err != nil {
+		return nil, errs.Wrapf(err, "unable to retrieve Auth token for '%s' service", forService)
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+
+	status := resp.StatusCode
+	if status != http.StatusOK {
+		log.Error(nil, map[string]interface{}{
+			"err":          err,
+			"request_path": authservice.ShowUserPath(),
+			"for_service":  forService,
+			"http_status":  status,
+		}, "failed to GET token from auth service due to HTTP error %s", status)
+		return nil, errs.Errorf("failed to GET Auth token for '%s' service due to status code %d", forService, status)
+	}
+
+	var respType authservice.TokenData
+	err = json.Unmarshal(respBody, &respType)
+	if err != nil {
+		log.Error(nil, map[string]interface{}{
+			"err":           err,
+			"request_path":  authservice.ShowUserPath(),
+			"for_service":   forService,
+			"http_status":   status,
+			"response_body": respBody,
+		}, "unable to unmarshal Auth token")
+		return nil, errs.Wrapf(err, "unable to unmarshal Auth token for '%s' service from Auth service", forService)
+	}
+
+	return respType.AccessToken, nil
+}
+
+// ensure tenantURLProvider implements BaseURLProvider
+var _ kubernetes.BaseURLProvider = &tenantURLProvider{}
+var _ kubernetes.BaseURLProvider = (*tenantURLProvider)(nil)
```

Comment:
just curious: what does this line do exactly ?

Reply:
I guess this make sure that tenantURLProvider struct always implements all method defined by kubernetes.BaseURLProvider interface.  If that not the case, compiler will complain.

Reply:
just to clarify, I know about the first line, but not about the second one.

Reply:
Honestly, I'm not sure why you need both (I'm too new to Go to work this out yet),  Best I can tell, Go treats pointers to objects and objects differently.  I was told to do it this way in a previous review, and indeed you'll see variants on these two lines throughout the WIT code.

Reply:
ok

Reply:
Hopefully I didn't suggested this code as it is here now because in both cases you're testing if the the `*tenentURLProvider` type implements the `kubernetes.BaseURLProvider` interface. What you want is to check that a the `tenentURLProvider` implements the interface with a value and with a pointer receiver. That is why the code should look like this:

```go
var _ kubernetes.BaseURLProvider = tenantURLProvider{} // NOTICE: absence of &
var _ kubernetes.BaseURLProvider = (*tenantURLProvider)(nil)
```

Reply:
Sorry, @kwk!  If you search for "var _" you'll see the whole sordid tale laid bare; both styles are in the WIT repo.  I am happy to change the code to remove the '&', but our previous discussion did include it.

I think the project needs a proper coding standards document.